### PR TITLE
feat: support sort agg

### LINKED ISSUE #591: executor: implement `SortAgg`
`SortAgg` is a special aggregator that takes an input of sorted group keys. For example:

```
select sum(x) from table group by group_key;

| group_key | x |
| a         | 1 |
| a         | 2 |
| a         | 3 |
| b         | 1 |
```

As data are already sorted with group_key, `SortAgg` can only have state for one group at a time, which significantly reduces memory footprint.

### PR DESCRIPTION
**Pull request by Kikkon**
close #591

### CODE REVIEW

**Comment by Kikkon** [comment]
please help review, and give me some more advice. 🤣

File: src/executor/sort_agg.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+                    for (state, expr) in states.iter_mut().zip_eq(&exprs) {
+                        state.update_single(&expr.get(row_idx))?;
+                    }
+                } else {
+                    let mut states = create_agg_states(&self.agg_calls);
+                    for (state, expr) in states.iter_mut().zip_eq(&exprs) {
+                        state.update_single(&expr.get(row_idx))?;
+                    }
+                    states_vec.push(states);
```
Comment (xxchan) [inline src/executor/sort_agg.rs:57]: Why not yield the agg result of last group here? If so, there's no need to store all previous states.

File: src/executor/sort_agg.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        };
+        let mut executor = executor.execute();
+
+        let expected_col = vec![1.3, 2.2, 0.3, 1.3, 0.4, 1.4, 0.5, 1.5];
+        let mut result_col = Vec::new();
+        while let Some(chunk) = executor.next().await {
+            for chunk in chunk.unwrap().arrays() {
+                result_col.push(chunk.clone());
+            }
+        }
```
Comment (xxchan) [inline src/executor/sort_agg.rs:126]: This test code seems a little bit confusing. Suggest to represent `expected` as `vec![vec![1.3, 2.2], ...]` And then assert each chunk looks like that.

File: src/executor/sort_agg.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+#![allow(dead_code)]
+use itertools::Itertools;
```
Comment (xxchan) [inline src/executor/sort_agg.rs:3]: Can it be removed?

File: src/executor/sort_agg.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    pub agg_calls: Vec<BoundAggCall>,
+    pub group_keys: Vec<BoundExpr>,
+    pub child: BoxedExecutor,
+}
+
+impl SortAggExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        let mut states_vec: Vec<SmallVec<[Box<dyn AggregationState>; 16]>> = vec![];
+        let mut last_key = Arc::new(HashKey::new());
```
Comment (xxchan) [inline src/executor/sort_agg.rs:21]: Is the `Arc` unnecessary?

File: src/executor/sort_agg.rs [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+            let num_rows = chunk.cardinality();
+            for row_idx in 0..num_rows {
+                // Create group key
+                let mut group_key = HashKey::new();
+                for col in group_cols.iter() {
+                    group_key.push(col.get(row_idx));
+                }
+                let group_key = Arc::new(group_key);
+                if last_key.len() != 0 && group_key.eq(&last_key) {
```
Comment (xxchan) [inline src/executor/sort_agg.rs:46]: I suggest to use `Option<HashKey>` instead of `last_key.len() != 0`

**Review by xxchan — COMMENTED** [review]
Please ping me again when you are ready for review. Also feel free to ask for help about specific questions. 😊

File: src/executor/sort_agg.rs (hunk 1 above)
Comment (Kikkon) [reply src/executor/sort_agg.rs]: When you have time, hope to help review again. I'm not familiar with async in rust, if there are mistakes, please help point them out (by the way, there are related materials that can also be recommended.  (：thx

File: src/executor/sort_agg.rs [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                .map(|agg| agg.args[0].eval(&chunk))
+                .try_collect()?;
+            let group_cols: SmallVec<[ArrayImpl; 16]> = self
+                .group_keys
+                .iter()
+                .map(|e| e.eval(&chunk))
+                .try_collect()?;
+
+            let num_rows = chunk.cardinality();
+            let mut states = create_agg_states(&self.agg_calls);
```
Comment (xxchan) [inline src/executor/sort_agg.rs:37]: Should this variable be defined outside the `for chunk in self.child` loop?

File: src/executor/sort_agg.rs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+            for row_idx in 0..num_rows {
+                // Create group key
+                let mut group_key = HashKey::new();
+                for col in group_cols.iter() {
+                    group_key.push(col.get(row_idx));
+                }
+                // Check group key & lask key
+                if last_key.is_some() && group_key.eq(last_key.as_ref().unwrap()) {
```
Comment (xxchan) [inline src/executor/sort_agg.rs:45]: `is_some` + `unwrap` (2 operations) can be simplified as `if let Some(..) = opt` or `match opt` (1 operation).

File: src/executor/sort_agg.rs (hunk 1 above)
Comment (xxchan) [reply src/executor/sort_agg.rs]: Be confident!

File: src/executor/sort_agg.rs (hunk 6 above)
Comment (xxchan) [reply src/executor/sort_agg.rs:37]: e.g., child may have two chunks `[1,1], [1,2]`

File: src/executor/sort_agg.rs [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+pub struct SortAggExecutor {
+    pub agg_calls: Vec<BoundAggCall>,
+    pub group_keys: Vec<BoundExpr>,
+    pub child: BoxedExecutor,
+}
+
+impl SortAggExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        let mut last_key = None::<SmallVec<[DataValue; 16]>>;
```
Comment (xxchan) [inline src/executor/sort_agg.rs:19]:
```suggestion
        let mut last_key = None::<HashKey>;
```

File: src/executor/sort_agg.rs [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+                        state.update_single(&expr.get(row_idx))?;
+                    }
+                } else {
+                    if last_key.is_some() {
+                        yield finish_agg(states);
+                        states = create_agg_states(&self.agg_calls);
+                    }
+                    for (state, expr) in states.iter_mut().zip_eq(&exprs) {
+                        state.update_single(&expr.get(row_idx))?;
+                    }
```
Comment (xxchan) [inline src/executor/sort_agg.rs:56]: Please consider again the branching logic here. Actually you only need one `if`, and then update state outside if

```rust 
if new agg key occurs {
    yield current state & reset state
}
update state acoording to the row
```

File: src/executor/sort_agg.rs (hunk 1 above)
Comment (xxchan) [reply src/executor/sort_agg.rs]: BTW, you don't need to be familiar with async actually. You can just "feel it". 😄

Specifically, to finish this work you just need to know the **semantics** of `yield` and generators.

File: src/executor/sort_agg.rs (hunk 6 above)
Comment (Kikkon) [reply src/executor/sort_agg.rs:37]: You mean [1,1] [1,2], group key in index[0], the result is [3,2]？

Comment (xxchan) [reply src/executor/sort_agg.rs]: Yes

**Review by xxchan — COMMENTED** [review]
Others LGTM! Ping @skyzh to take a look.

File: src/executor/sort_agg.rs [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+                        }
+                    }
+                    None => (),
+                }
+                for (state, expr) in states.iter_mut().zip_eq(&exprs) {
+                    state.update_single(&expr.get(row_idx))?;
+                }
+                last_key = Some(group_key);
+            }
+            yield finish_agg(&states)
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/executor/sort_agg.rs:59]: So I think adding relevant tests may be better..

**Comment by xxchan** [comment]
BTW, since we will squash merge, there's no need to force push everytime to make commit history (too) clean. Leave changes in new commits may also help reviewers to focus on new changes since last review 😄

**Review by skyzh — COMMENTED** [review]
Would you please add a test case when there are no rows from child? We should output no rows instead of one row of all null. I guess the current behavior is incorrect 🤣

File: src/executor/sort_agg.rs [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+        for index in agg_call_index {
+            agg_calls.push(create_sum_agg_call(index));
+        }
+
+        let mut group_keys = vec![];
+        for index in group_key_index {
+            group_keys.push(create_input_ref(index));
+        }
+
+        let child: BoxedExecutor = async_stream::try_stream! {
```
[the remark this answers is not in the document]
Comment (Kikkon) [reply src/executor/sort_agg.rs:186]: SimpleAggExecutor has the same problem too?

[the remark this answers is not in the document]
Comment (Kikkon) [reply src/executor/sort_agg.rs:193]: get it, thx!

**Comment by Kikkon** [comment]
@xxchan @skyzh  any other suggestions? 😄

**Comment by skyzh** [comment]
Let me take a look 🤣

File: src/executor/sort_agg.rs [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+                .try_collect()?;
+
+            let num_rows = chunk.cardinality();
+            for row_idx in 0..num_rows {
+                // Create group key
+                let mut group_key = HashKey::new();
+                for col in group_cols.iter() {
+                    group_key.push(col.get(row_idx));
+                }
+                // Check group key & lask key
```
Comment (skyzh) [inline src/executor/sort_agg.rs:44]: lask -> last

File: src/executor/sort_agg.rs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+                .iter()
+                .map(|e| e.eval(&chunk))
+                .try_collect()?;
+
+            let num_rows = chunk.cardinality();
+            for row_idx in 0..num_rows {
+                // Create group key
+                let mut group_key = HashKey::new();
+                for col in group_cols.iter() {
+                    group_key.push(col.get(row_idx));
```
Comment (skyzh) [inline src/executor/sort_agg.rs:42]: Maybe we can optimize this in the future. We can first check group key is the same as the current group key, and then clone by `.get` if different. Otherwise we will always need to create new vector for every row of hash key.

**Review by skyzh — APPROVED** [review]
Rest LGTM, thanks! Would you please help fix the typo?

File: src/executor/sort_agg.rs (hunk 12 above)
Comment (Kikkon) [reply src/executor/sort_agg.rs]: sure.

**Comment by Kikkon** [comment]
Thank you for your support 😊  @xxchan @skyzh
